File: packages/pdf-gen-js/src/utils/table.ts
import { PDFConfig } from '../config';
import { Theme } from '../theme';
import { PDFDoc } from '../types';

export interface TableColumn {
  header: string;
  key?: string;
  width?: number;
  align?: 'left' | 'center' | 'right';
}

export interface TableOptions {
  rowHeight?: number;
  headerHeight?: number;
  striped?: boolean;
  fontSize?: number;
  startY?: number;
}

/**
 * Draws a table with word-wrapped cells, an optional striped body, and automatic
 * page breaks when a row would run past the bottom margin. Returns the Y position
 * immediately below the finished table.
 */
export function drawTable(
  doc: PDFDoc,
  config: PDFConfig,
  theme: Theme,
  columns: TableColumn[],
  rows: Array<Record<string, any> | any[]>,
  options: TableOptions = {}
): number {
  const x = config.margins.left;
  const totalWidth = doc.page.width - config.margins.left - config.margins.right;
  const headerHeight = options.headerHeight ?? 26;
  const fontSize = options.fontSize ?? 10;
  const striped = options.striped ?? true;

  const explicitWidth = columns.reduce((sum, c) => sum + (c.width || 0), 0);
  const flexColumns = columns.filter((c) => !c.width).length;
  const flexWidth = flexColumns > 0 ? (totalWidth - explicitWidth) / flexColumns : 0;
  const widths = columns.map((c) => c.width || flexWidth);

  const cellValue = (row: Record<string, any> | any[], col: TableColumn, idx: number): string => {
    const raw = Array.isArray(row) ? row[idx] : row[col.key as string];
    return raw === undefined || raw === null ? '' : String(raw);
  };

  const rowLineCount = (row: Record<string, any> | any[]): number => {
    let maxLines = 1;
    columns.forEach((col, i) => {
      const text = cellValue(row, col, i);
      const height = doc.heightOfString(text, { width: widths[i] - 12 });
      const lines = Math.ceil(height / (fontSize * 1.3));
      maxLines = Math.max(maxLines, lines || 1);
    });
    return maxLines;
  };

  const drawHeader = (y: number): number => {
    doc.rect(x, y, totalWidth, headerHeight).fill(theme.tableHeaderBg);
    doc.fillColor(theme.tableHeaderText).font('Helvetica-Bold').fontSize(fontSize);
    let colX = x;
    columns.forEach((col, i) => {
      doc.text(col.header, colX + 6, y + headerHeight / 2 - fontSize / 2, {
        width: widths[i] - 12,
        align: col.align || 'left'
      });
      colX += widths[i];
    });
    return y + headerHeight;
  };

  let y = options.startY ?? doc.y;
  if (y + headerHeight > doc.page.height - config.margins.bottom) {
    doc.addPage();
    y = config.margins.top;
  }
  y = drawHeader(y);

  rows.forEach((row, rowIdx) => {
    const lines = rowLineCount(row);
    const rowHeight = Math.max(options.rowHeight ?? 22, lines * fontSize * 1.3 + 10);

    if (y + rowHeight > doc.page.height - config.margins.bottom) {
      doc.addPage();
      y = drawHeader(config.margins.top);
    }

    if (striped && rowIdx % 2 === 1) {
      doc.rect(x, y, totalWidth, rowHeight).fill(theme.tableStripeBg);
    }

    doc.fillColor(theme.text).font('Helvetica').fontSize(fontSize);
    let colX = x;
    columns.forEach((col, i) => {
      doc.text(cellValue(row, col, i), colX + 6, y + 6, {
        width: widths[i] - 12,
        align: col.align || 'left'
      });
      colX += widths[i];
    });

    doc
      .strokeColor(theme.border)
      .lineWidth(0.5)
      .moveTo(x, y + rowHeight)
      .lineTo(x + totalWidth, y + rowHeight)
      .stroke();

    y += rowHeight;
  });

  doc.x = x;
  doc.y = y;
  return y;
}

File: packages/pdf-gen-py/pdf_gen/utils/table.py
from typing import Any, Dict, List, Optional, Union

from ..config import PDFConfig, Theme
from ..surface import Surface

Row = Union[Dict[str, Any], List[Any]]


def _cell_value(row: Row, col: Dict[str, Any], idx: int) -> str:
    raw = row[idx] if isinstance(row, list) else row.get(col.get('key'))
    return '' if raw is None else str(raw)


def draw_table(
    surface: Surface,
    config: PDFConfig,
    theme: Theme,
    columns: List[Dict[str, Any]],
    rows: List[Row],
    start_y: float,
    font_size: float = 10,
    header_height: float = 26,
    row_height: float = 22,
    striped: bool = True,
) -> float:
    """Draws a table using top-down coordinates, word-wrapping cells and
    breaking to a new page when a row would run past the bottom margin.
    Returns the y (top-down) position immediately below the table."""

    x = config.margins['left']
    total_width = surface.page_width - config.margins['left'] - config.margins['right']

    explicit_width = sum(c.get('width') or 0 for c in columns)
    flex_columns = sum(1 for c in columns if not c.get('width'))
    flex_width = (total_width - explicit_width) / flex_columns if flex_columns else 0
    widths = [c.get('width') or flex_width for c in columns]

    def draw_header(y: float) -> float:
        surface.rect(x, y, total_width, header_height, fill=theme.table_header_bg)
        col_x = x
        for i, col in enumerate(columns):
            surface.text(
                col_x + 6,
                y + header_height / 2 - font_size / 2,
                col['header'],
                font='Helvetica-Bold',
                size=font_size,
                color=theme.table_header_text,
                align=col.get('align', 'left'),
                width=widths[i] - 12,
            )
            col_x += widths[i]
        return y + header_height

    def row_lines_height(row: Row) -> float:
        max_h = row_height
        for i, col in enumerate(columns):
            text = _cell_value(row, col, i)
            h = surface.measure_height(text, 'Helvetica', font_size, widths[i] - 12) + 10
            max_h = max(max_h, h)
        return max_h

    bottom_limit = surface.page_height - config.margins['bottom']

    y = start_y
    if y + header_height > bottom_limit:
        surface.new_page()
        y = config.margins['top']
    y = draw_header(y)

    for row_idx, row in enumerate(rows):
        height = row_lines_height(row)

        if y + height > bottom_limit:
            surface.new_page()
            y = config.margins['top']
            y = draw_header(y)

        if striped and row_idx % 2 == 1:
            surface.rect(x, y, total_width, height, fill=theme.table_stripe_bg)

        col_x = x
        for i, col in enumerate(columns):
            surface.text(
                col_x + 6,
                y + 6,
                _cell_value(row, col, i),
                font='Helvetica',
                size=font_size,
                color=theme.text,
                align=col.get('align', 'left'),
                width=widths[i] - 12,
            )
            col_x += widths[i]

        surface.line(x, y + height, x + total_width, y + height, color=theme.border, width=0.5)
        y += height

    return y
